material layering part 2: layers + splatmaps

HifiExperiments opened the pull request:
diff here: https://github.com/HifiExperiments/overte/compare/materialLayers...HifiExperiments:overte:materialLayers2

follow up to #1531, this is purely the shader changes to fully implement layering + splat mapping.  the variants are still not enabled, so this should have no functional change yet

## Funding

This project is funded through [NGI Zero Core](https://nlnet.nl/core), a fund established by [NLnet](https://nlnet.nl) with financial support from the European Commission's [Next Generation Internet](https://ngi.eu) program. Learn more at the [NLnet project page](https://nlnet.nl/project/Overte-visualscripting).

[<img src="https://nlnet.nl/logo/banner.png" alt="NLnet foundation logo" width="20%" />](https://nlnet.nl)
[<img src="https://nlnet.nl/image/logos/NGI0_tag.svg" alt="NGI Zero Logo" width="20%" />](https://nlnet.nl/core)

HifiExperiments commented:
@JulianGro could you give this a check for the gles shaders?

ksuprynowicz commented:
The breakage happens with both deferred and forward renderers.

ksuprynowicz commented:
The issue seems to be fixed fine if I comment out `discard` statement in shader in RederDoc:
```
}
            fragNormalWSLayer = evalFrontOrBackFaceNormal(normalize(fragNormalWSLayer));

        float magnitude = layerBlend[i];
            albedo += albedoLayer * magnitude;
            opacity += opacityLayer * magnitude;
            fragNormalWS += fragNormalWSLayer * magnitude;
            roughness += roughnessLayer * magnitude;
            metallic += metallicLayer * magnitude;
            emissive += emissiveLayer * magnitude;
                occlusion += occlusionLayer * magnitude;
                scattering += scatteringLayer * magnitude;
                }

    albedo *= _color.rgb;

    {
    /*if (opacity < 1.0) {
        discard;
    }*/
}

    TransformCamera cam = getTransformCamera();
        vec3 fresnel = getFresnelF0(metallic, albedo);
        vec4 color = vec4(evalSkyboxGlobalColor(
                    cam._viewInverse,
                    1.0,
                    occlusion,
                    _positionES.xyz,
                    fragNormalWS,
                    albedo,
                    fresnel,
                    metallic,
                    roughness),
                    opacity);
                color.rgb += emissive * isEmissiveEnabled();
                _fragColor0 = color;
            }
```

ksuprynowicz commented:
Changing the opacity to something lower than 1.0 also fixes it:
```
    if (opacity < 0.5) {
        discard;
    }
```

HifiExperiments commented:
oooo interesting, what about < 0.99999 or something like that?

ksuprynowicz commented:
> oooo interesting, what about < 0.99999 or something like that?

I just checked and 0.99999 does the job too :)

ksuprynowicz commented:
I also noticed one more possible issue, but it might have happened earlier.
It seems like shaders loaded in-game are not optimized with SPIR-V optmizer anymore? If so, this could lead to longer shader compilation time.

ksuprynowicz commented:
Could we store shaders outside of qrc, for example in some directory?

HifiExperiments commented:
I think that’s more or less what the qrc stuff does.  the problem is that we have so many shader variants now that the compiler runs out of heap space when trying to include them.  supposedly splitting them into multiple qrc files would help but that requires some other changes I haven’t worked out yet.  as soon as we get this and #1592 merged I will be able to look at it again

HifiExperiments commented:
@ksuprynowicz merge conflicts resolved and updated the opacity cutoff 👍

ksuprynowicz commented:
It looks like the issue is still happening, and also transparencies and bloom seem broken on voxel entities. There are various kinds of PBR voxels in Overte Hub that can be used for testing:
![overte-snap-by--on-2025-07-06_13-01-40](https://github.com/user-attachments/assets/9f6e1e1c-de4e-4664-892a-feeb55942d84)

ksuprynowicz commented:
The current version also runs only once for me, then crashes on every consecutive run. Here's the backtrace:
```
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at ./nptl/pthread_kill.c:44
#1  0x00007ffff209e9ff in __pthread_kill_internal (threadid=<optimized out>, signo=6) at ./nptl/pthread_kill.c:89
#2  0x00007ffff2049cc2 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#3  0x00007ffff20324ac in __GI_abort () at ./stdlib/abort.c:73
#4  0x00007ffff22a1a3d in __gnu_cxx::__verbose_terminate_handler () at ../../../../src/libstdc++-v3/libsupc++/vterminate.cc:95
#5  0x00007ffff22b344a in __cxxabiv1::__terminate (handler=<optimized out>) at ../../../../src/libstdc++-v3/libsupc++/eh_terminate.cc:48
#6  0x00007ffff22a15e9 in std::terminate () at ../../../../src/libstdc++-v3/libsupc++/eh_terminate.cc:58
#7  0x00007ffff22b36c8 in __cxxabiv1::__cxa_throw (obj=<optimized out>, tinfo=0x7ffff2455bc0 <typeinfo for std::bad_alloc>, dest=0x7ffff22b1950 <std::bad_alloc::~bad_alloc()>) at ../../../../src/libstdc++-v3/libsupc++/eh_throw.cc:98
#8  0x00007ffff6c91341 in qBadAlloc() () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#9  0x00007ffff6c95ea3 in ??? () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#10 0x00007ffff6f3e0fc in ??? () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#11 0x00007ffff6d6ac4d in QString::toUtf8_helper(QString const&) () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#12 0x00007ffff54f5cab in QString::toUtf8() const & (this=0x7fffffffaf78) at /usr/include/x86_64-linux-gnu/qt5/QtCore/qstring.h:681
#13 gl::loadShaderCache (cache=std::unordered_map with 0 elements) at /home/ksuprynowicz/overte/overte_v8/overte/libraries/gl/src/gl/GLShaders.cpp:445
#14 0x00007ffff6b8462a in gpu::gl::GLBackend::initShaderBinaryCache (this=this@entry=0x55555a50c3f0) at /home/ksuprynowicz/overte/overte_v8/overte/libraries/gpu-gl-common/src/gpu/gl/GLBackendShader.cpp:400
#15 0x00007ffff6b7c43e in gpu::gl::GLBackend::GLBackend (this=this@entry=0x55555a50c3f0) at /home/ksuprynowicz/overte/overte_v8/overte/libraries/gpu-gl-common/src/gpu/gl/GLBackend.cpp:282
#16 0x00007ffff7e8ab8f in gpu::gl45::GL45Backend::GL45Backend (this=this@entry=0x55555a50c3f0) at /home/ksuprynowicz/overte/overte_v8/overte/libraries/gpu-gl/src/gpu/gl45/GL45Backend.cpp:57
#17 0x00007ffff7e7fc97 in std::_Construct<gpu::gl45::GL45Backend> (__p=0x55555a50c3f0) at /usr/include/c++/14/bits/stl_construct.h:109
#18 std::allocator_traits<std::allocator<void> >::construct<gpu::gl45::GL45Backend> (__p=0x55555a50c3f0) at /usr/include/c++/14/bits/alloc_traits.h:706
#19 std::_Sp_counted_ptr_inplace<gpu::gl45::GL45Backend, std::allocator<void>, (__gnu_cxx::_Lock_policy)2>::_Sp_counted_ptr_inplace<>(std::allocator<void>) (this=0x55555a50c3e0, __a=...) at /usr/include/c++/14/bits/shared_ptr_base.h:607
#20 std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<gpu::gl45::GL45Backend, std::allocator<void>>(gpu::gl45::GL45Backend*&, std::_Sp_alloc_shared_tag<std::allocator<void> >) (this=<optimized out>, __p=<optimized out>, __a=...)
    at /usr/include/c++/14/bits/shared_ptr_base.h:969
#21 std::__shared_ptr<gpu::gl45::GL45Backend, (__gnu_cxx::_Lock_policy)2>::__shared_ptr<std::allocator<void>>(std::_Sp_alloc_shared_tag<std::allocator<void> >) (this=<optimized out>, __tag=...) at /usr/include/c++/14/bits/shared_ptr_base.h:1713
#22 std::shared_ptr<gpu::gl45::GL45Backend>::shared_ptr<std::allocator<void>>(std::_Sp_alloc_shared_tag<std::allocator<void> >) (this=<optimized out>, __tag=...) at /usr/include/c++/14/bits/shared_ptr.h:463
#23 std::make_shared<gpu::gl45::GL45Backend> () at /usr/include/c++/14/bits/shared_ptr.h:1008
#24 gpu::gl::GLBackend::createBackend () at /home/ksuprynowicz/overte/overte_v8/overte/libraries/gpu-gl/src/gpu/gl/GLDesktopBackend.cpp:37
#25 0x00007ffff61e6610 in gpu::Context::Context (this=this@entry=0x55555a2b1650) at /home/ksuprynowicz/overte/overte_v8/overte/libraries/gpu/src/gpu/Context.cpp:53
#26 0x0000555555979b4f in std::_Construct<gpu::Context> (__p=0x55555a2b1650) at /usr/include/c++/14/bits/stl_construct.h:109
#27 std::allocator_traits<std::allocator<void> >::construct<gpu::Context> (__p=0x55555a2b1650) at /usr/include/c++/14/bits/alloc_traits.h:706
#28 std::_Sp_counted_ptr_inplace<gpu::Context, std::allocator<void>, (__gnu_cxx::_Lock_policy)2>::_Sp_counted_ptr_inplace<>(std::allocator<void>) (this=0x55555a2b1640, __a=...) at /usr/include/c++/14/bits/shared_ptr_base.h:607
#29 std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<gpu::Context, std::allocator<void>>(gpu::Context*&, std::_Sp_alloc_shared_tag<std::allocator<void> >) (this=<optimized out>, __p=<optimized out>, __a=...)
    at /usr/include/c++/14/bits/shared_ptr_base.h:969
#30 std::__shared_ptr<gpu::Context, (__gnu_cxx::_Lock_policy)2>::__shared_ptr<std::allocator<void>>(std::_Sp_alloc_shared_tag<std::allocator<void> >) (this=<optimized out>, __tag=...) at /usr/include/c++/14/bits/shared_ptr_base.h:1713
#31 std::shared_ptr<gpu::Context>::shared_ptr<std::allocator<void>>(std::_Sp_alloc_shared_tag<std::allocator<void> >) (this=<optimized out>, __tag=...) at /usr/include/c++/14/bits/shared_ptr.h:463
#32 std::make_shared<gpu::Context> () at /usr/include/c++/14/bits/shared_ptr.h:1008
#33 GraphicsEngine::initializeGPU (this=0x555559c26ce0, glwidget=<optimized out>) at /home/ksuprynowicz/overte/overte_v8/overte/interface/src/graphics/GraphicsEngine.cpp:59
#34 0x000055555584eb31 in Application::initializeGL (this=this@entry=0x7fffffffc120) at /usr/include/c++/14/bits/shared_ptr_base.h:1666
#35 0x0000555555870233 in Application::initialize (this=this@entry=0x7fffffffc120, parser=...) at /home/ksuprynowicz/overte/overte_v8/overte/interface/src/Application_Setup.cpp:792
#36 0x0000555555726048 in main (argc=<optimized out>, argv=<optimized out>) at /home/ksuprynowicz/overte/overte_v8/overte/interface/src/main.cpp:753
```

HifiExperiments commented:
@ksuprynowicz I've updated the alpha cutoff in more places and fixed the forward/transparent rendering which I believe has fixed the bloom issue also.  I have some more clues to fixing the larger bloom bug but will save that for a separate PR

ksuprynowicz commented:
Thank you!
I tested it and now everything renders perfectly :)
One time during testing shader cache got corrupted and then game couldn't start at all anymore, I saved the corrupted cache json file so we can figure out why it causes a crash later.
